File: 150.evaluate-reverse-polish-notation/solution.kt
import java.util.Stack

class Solution {
    fun calculate(val1: Int, action: String, val2: Int): String {
        return (when(action) {
            "+" -> val1 + val2
            "*" -> val1 * val2
            "-" -> val1 - val2
            "/" -> val1 / val2
            else -> "0"
        }).toString()
    }

    fun evalRPN(tokens: Array<String>): Int {
        val stack = Stack<String>()
        for (token in tokens) {
            stack.push(when(token) {
                "+", "-", "*", "/" -> {
                    val val2 = stack.pop().toInt()
                    val val1 = stack.pop().toInt()
                    this.calculate(val1, token, val2)
                }

                else -> token
            })
        }

        var result = 0
        for (token in stack) {
            result += token.toInt()
        }
        return result
    }
}

File: 150.evaluate-reverse-polish-notation/solution.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        actions = {
            "+": lambda x, y: x + y,
            "-": lambda x, y: x - y,
            "*": lambda x, y: x * y,
            "/": lambda x, y: int(x / y)
        }
        stack = []

        for t in tokens:
            if (not stack):
                stack.append(t)
                continue

            if (t not in actions):
                stack.append(t)
                continue

            val2 = int(stack.pop())
            val1 = int(stack.pop())
            value = actions[t](val1, val2)
            # print(val1, t, val2, "->", value)
            stack.append(str(value))

        result = 0
        for t in stack:
            result += int(t)
        return result
